design/proposals/proxy-support/proxy-support.md

```diff
@@ -0,0 +1,157 @@
+# Maintaining Requester IP Addresses <!-- omit in toc --> 
+
+When Conjur (or DAP) are run behind a load balancer, users are unable to use CIDR restrictions on Hosts or Layers. Furthermore, Load Balancer IP address appears in Audit logs instead of the true client IP address. This is because Conjur references the last entry in the `X-Forwarded-For` header.
+
+This decision was conscious, and meant to prevent IP spoofing. We have reached a point where customers need a mechanism to add known proxies to Conjur/DAP to identify the true identity of the requesting client. 
+
+This design document is meant to provide a possible solution to the challenge of identifying the client IP while preventing IP spoofing.
+
+- [Current Request Flow](#current-request-flow)
+- [Proposed Solution (Summary)](#proposed-solution-summary)
+- [Proposed Solution (Detailed)](#proposed-solution-detailed)
+  - [Adding Proxies with Policy](#adding-proxies-with-policy)
+  - [Limitations](#limitations)
+    - [Proxies cannot be layered](#proxies-cannot-be-layered)
+- [Rejected Alternatives](#rejected-alternatives)
+  - [Configuration using Environment Variables](#configuration-using-environment-variables)
+
+## Current Request Flow
+
+The following visualization offers a typical experience as a requests moves through a Layer 7 load balancers.
+
+![](/design/diagrams/out/proxy-support-overview/proxy-support-overview.png)
+
+As we can see, each load balancer adds its IP address to the `X-Forwarded-For` header.
+
+With our current implementation:
+```ruby
+# Rack::Request
+def ip
+  if addr = @env['HTTP_X_FORWARDED_FOR']
+    addr.split(',').last.strip
+  else
+    @env['REMOTE_ADDR']
+  end
+end
+```
+the IP address of the last load balancer (`10.2.0.1`) would be used as the request IP.
+
+## Proposed Solution (Summary)
+Users can define the proxies in front of a node(s) using Conjur Policy. This enables Conjur to use the first non-proxy IP as the client IP.
+
+## Proposed Solution (Detailed)
+
+### Adding Proxies with Policy
+
+Given the following setup
+![](/design/diagrams/out/proxy-master-node/proxy-master-node.png)
+
+we can apply the following policy:
+```yaml
+- !policy
+  id: conjur/settings/proxies
+  body:
+    - !host 
+      id: dap-master.mycompany.com
+      annotations:
+        proxy/ip-addresses: ['1.2.3.4']
```

Comment:
Will it support CIDR notation?
I suppose that it's a possible configuration to have a pool of load balancers.

Reply:
Yes. I didn't include it because I'm not sure how common it would be for all load balancers to be in the same network space.  I've always set them up in an application of environment subnet. In that case, a single IP would be fine.